### ISSUE TITLE
feat: implement Signal type for reactive programming

### ISSUE THREAD

**Issue by KentaroMorishita** [KentaroMorishita/seseragi#83]
# Signal型実装 - リアクティブプログラミング機能

## 概要

RBoxの実証済み設計思想を採用し、Seseragi言語にリアクティブプログラミングのSignal型を実装する。既存のモナド演算子（Task, Maybe, Either）と完全統合し、型安全なミュータブル値を提供する。

### 基本操作

```seseragi
// Signal作成（型コンストラクタ）
let count = Signal 0                           // Signal<Int>
let message = Signal "hello"                   // Signal<String>
let user = Signal (User { name: "Alice", age: 25 })  // Signal<User>
```

#### 1. 直接値設定
```seseragi
count := 5          // count.setValue(5) → 値は5
message := "world"  // message.setValue("world") → 値は"world"
```

#### 2. ラムダ関数による前値ベース更新  
```seseragi
count := (\x -> x + 1)        // インクリメント：前の値 + 1
count := (\prev -> prev * 2)  // ダブル：前の値 * 2
message := (\s -> s + "!")    // 文字列追加：前の値 + "!"
```

#### 3. 名前付き関数による更新
```seseragi
fn increment prev: Int -> Int = prev + 1
fn double prev: Int -> Int = prev * 2
fn addExclamation prev: String -> String = prev + "!"

count := increment      // count.setValue(increment)
count := double        // count.setValue(double) 
message := addExclamation  // message.setValue(addExclamation)
```

#### 4. 部分適用関数による更新
```seseragi
fn add x: Int -> y: Int -> Int = x + y
let addFive = add 5

count := addFive    // count.setValue(addFive) → prev + 5
```

#### Functor (`<$>`)
```seseragi
let count = Signal 10
let doubled = (\x -> x * 2) <$> count   // Signal<Int>: countの変更に自動反応
let formatted = (\x -> "Count: " + toString x) <$> count  // Signal<String>
```

#### Applicative (`<*>`)
```seseragi
let x = Signal 10
let y = Signal 5
let sum = Just (\a b -> a + b) <*> x <*> y  // Signal<Int>: x,y変更時自動再計算
```

#### Monad (`>>=`)
```seseragi
let userId = Signal 123
let userProfile = userId >>= (\id -> fetchUserProfile id)  // Signal<Profile>
// userIdが変更されると新しいプロファイルを自動取得
```

### サブスクライブ機能（明示的制御）

```seseragi
// Signal変更の監視
let sub = subscribe count (\x -> print ("Count changed: " + toString x))

// 購読解除
unsubscribe sub

// 複数サブスクライバー
let sub1 = subscribe count (\x -> print ("Sub1: " + toString x))
let sub2 = subscribe count (\x -> log x)
```

### 個別detach設計（選択的依存切断）

```seseragi
let root = Signal 10                    // レベル0
let child1 = (\x -> x * 2) <$> root    // レベル1: rootに依存（値: 20）
let child2 = (\x -> x + 1) <$> child1  // レベル2: child1に依存（値: 21）

// 通常の動作
root := 5   // root=5, child1=10, child2=11 (全連動)

// 選択的切断
child1.detach()  // child1だけrootから独立

// 切断後の動作
root := 100     // root=100, child1=10(変化なし), child2=11(変化なし)
child1 := 50    // root=100, child1=50, child2=51(child1の変更には反応)
```

### detachHandlers仕組み
- 各Signalが自分の購読解除関数を `detachHandlers` 配列で管理
- 親子双方向参照なし（循環参照回避）
- ねずみ講的伝搬なし（個別制御可能）

### Signal<T>型の位置づけ
- Seseragi唯一のミュータブル型
- Task, Maybe, Either と同列の型として扱う
- 型パラメータTは不変（値レベルでのみ変更可能）

### `:=` 演算子の型チェック
```seseragi
let count = Signal 0  // Signal<Int>

// 型安全な更新
count := 5              // ✅ Int型
count := (\x -> x + 1)  // ✅ Int -> Int型
count := "hello"        // ❌ 型エラー: String は Int に代入不可
```

### Phase 1: Signal基本実装 (`src/runtime/seseragi-runtime.ts`)
```typescript
// Signal型定義（RBoxのdetachHandlers採用）
export type Signal<T> = {
  readonly isSignal: true
  readonly getValue: () => T
  readonly setValue: ((value: T) => void) & ((value: (prev: T) => T) => void)
  readonly detach: () => void
  readonly detachHandlers: (() => void)[]
}

// Signal専用ランタイム関数（prefix統一）
export const createSignal = <T>(initialValue: T): Signal<T>
export const setSignal = <T>(signal: Signal<T>, value: T | ((prev: T) => T)): void
export const subscribeSignal = <T>(signal: Signal<T>, observer: (value: T) => void): string
export const unsubscribeSignal = <T>(signal: Signal<T>, key: string): void
export const detachSignal = <T>(signal: Signal<T>): void
```

### Phase 2: モナド演算子実装（用語整理版）
```typescript
// Signal専用モナド操作
export const mapSignal = <T, U>(sourceSignal: Signal<T>, f: (value: T) => U): Signal<U>
export const applySignal = <T, U>(functionSignal: Signal<(a: T) => U>, valueSignal: Signal<T>): Signal<U>
export const bindSignal = <T, U>(sourceSignal: Signal<T>, f: (value: T) => Signal<U>): Signal<U>
```

### Phase 3: 構文・型システム統合
- `src/ast.ts`: SignalType, SignalExpression, AssignmentExpression追加
- `src/parser.ts`: Signal構文、`:=`演算子（文脈判断で既存コロン構文と区別）
- `src/type-inference.ts`: Signal<T>型推論、`:=`の `T | (T -> T)` 型チェック

### Phase 4: コード生成
- `src/codegen.ts`: TypeScript Signal クラス生成
- リアクティブ更新メカニズムとcleanup処理

### Phase 5: VS Code統合・テスト
- 構文ハイライト更新
- 包括的テストケース（RBox機能と同等）

### 基本機能テスト
```seseragi
// Signal作成・更新
let signal = Signal 42
signal := 100
signal := (\x -> x * 2)

// モナド演算
let mapped = (\x -> x + 1) <$> signal
let applied = Just (\x y -> x + y) <*> signal <*> Signal 10
let bound = signal >>= (\x -> Signal (x * x))
```

### detach機能テスト
```seseragi
let parent = Signal 10
let child = (\x -> x * 2) <$> parent
child.detach()
// parentの変更がchildに影響しないことを確認
```

### エラーケーステスト
```seseragi
let intSignal = Signal 42
intSignal := "hello"  // 型エラーが発生することを確認
```

## 期待される成果

この実装により、Seseragiは以下を実現する：

1. **Task（非同期処理）** + **Signal（リアクティブプログラミング）** を統合した現代的関数型言語
2. 型安全なミュータブル値による実用性向上
3. RBoxの実証済み設計による安定性
4. 既存モナド演算子との完全統合による学習コストミニマム化
5. 選択的detachによる柔軟なリソース管理

## 参考資料

- [f-box-core RBox implementation](https://github.com/KentaroMorishita/f-box-core/blob/main/src/RBox.ts)
- Seseragi既存モナド実装 (`src/runtime/seseragi-runtime.ts`)

---

**Priority**: High  
**Labels**: enhancement, reactive-programming, signal  
**Milestone**: Signal型実装
